2016/exercises/final_project/Gabriel-Adrian/main.cpp

```diff
@@ -0,0 +1,519 @@
+// Enumeration of simplicial polytopes and neighbourly
+// simplicial polytopes in R⁴ with 8 vertices.
+//
+// Authors: Gabriel Riera Roca and Adrián Ranea Robles.
+//
+// Description:
+//   This program lists 37 simplicial polytopes and 3
+//   neighbourly simplicial polytopes in R⁴ with 8 vertices
+//   by generating random affine Gale diagrams.
+//
+//   For example, with the seed 1234567890, approximately
+//   32000 random diagrams are needed to list 3 neighbourly
+//   simplicial polytopes and 37 simplicial polytopes.
+//   This takes a couple of minutes in a personal computer.
+//
+//   Our list of polytopes agrees with the list obtained by
+//   Grünbaum in his paper “An enumeration of simplicial
+//   4-polytopes with 8 vertices”.
+//
+// Dependencies:
+//   GNU Multiple Precision Arithmetic Library
+//   Parma Polyhedra Library
+//
+#include <cstdint>
+#include <array>
+#include <vector>
+
+#include <gmpxx.h>
+#include <ppl.hh>
+namespace ppl = Parma_Polyhedra_Library;
+
+constexpr auto MAX_ITER = 1000000;
+
+// Basis for R²
+const auto X = ppl::Variable(0);
+const auto Y = ppl::Variable(1);
+
+// A point in R²
+struct Point
+{
+    mpq_class x, y;
+
+    Point(mpq_class x, mpq_class y) : x(x), y(y) {}
+
+    // Return a generator to be used with PPL (convenience)
+    ppl::Generator generator() const
+    {
+        auto nx = x.get_num();
+        auto dx = x.get_den();
+        auto ny = y.get_num();
+        auto dy = y.get_den();
+
+        mpz_class z = dx*dy;
+        return ppl::Generator::point(nx*dy*X + ny*dx*Y, z);
+    }
+};
+
+// A point in a Gale diagram
+struct GalePt
+{
+    enum Sign { Positive, Negative };
+
+    Point pt;
+    Sign sign;
+};
+
+// Representation of a Gale diagram
+struct GaleDiagram
+{
+    std::vector<GalePt> points;
+};
+
+// Stores the facets of a polytope.
+struct CombinatorialType
+{
+    // Compact representation of a subset of a given set of size tsize
+    // tsize is never larger than 8 in our case
+    struct PointSet
+    {
+        using Bits = std::uint32_t;
+        Bits bits;
+        size_t tsize; // total size
+
+        // Check for inclusion
+        bool contained_in(CombinatorialType::PointSet other) const
+        {
+            return (bits & other.bits) == bits;
+        }
+
+        // Test for membership of an element
+        bool contains(size_t i) const
+        {
+            return bits & (1 << i);
+        }
+
+        // Lexicographical ordering for sorting, etc
+        struct TotalOrder
+        {
+            bool operator()(PointSet a, PointSet b) const { return a.bits < b.bits; }
+        };
+    };
+
+    using VertId = size_t;
+    using Facet = PointSet;
+
+    size_t n_vert;
+    std::vector<Facet> facets; // sorted by TotalOrder
+};
+
+bool operator==(CombinatorialType::PointSet a, CombinatorialType::PointSet b)
+{
```

Comment:
Possibly use `const CombinatorialType::PointSet&` just to be explicit... maybe the compiler will do this anyway, but in any case `a` and `b` will still be valid after this function, so the compiler can't move from them